### PR TITLE
[TFL] Convert Size to Prod(Shape)

### LINKED ISSUE #46285: 'tf.Size' op is neither a custom op nor a flex op
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: No
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10.0.18363
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**: n/a
-   **TensorFlow installed from (source or binary)**: binary
-   **TensorFlow version (use command below)**: tensorflow-2.4.0
-   **Python version**: 3.8.5
-   **Bazel version (if compiling from source)**: n/a
-   **GCC/Compiler version (if compiling from source)**: n/a
-   **CUDA/cuDNN version**: CUDA 11.0 / CuDNN 8.0.5
-   **GPU model and memory**: Quadro P2000 4GB
-   **Exact command to reproduce**: Run "Convert a SavedModel" from https://www.tensorflow.org/lite/convert

### Describe the problem
For some unknown reason (to me) I am no longer able to convert a model to TFLite for usage on a Raspberry Pi. When I run the script mentioned in the link above, the following error appears:

```
loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): error: 'tf.Size' op is neither a custom op nor a flex op
error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
Traceback (most recent call last):
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 210, in toco_convert_protos
    model_str = wrap_toco.wrapped_toco_convert(model_flags_str,
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\wrap_toco.py", line 32, in wrapped_toco_convert
    return _pywrap_toco_api.TocoConvert(
Exception: <unknown>:0: error: loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): 'tf.Size' op is neither a custom op nor a flex op
<unknown>:0: note: loc("StatefulPartitionedCall"): called from
<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
 
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "convert_to_tflite_v2.py", line 7, in <module>
    tflite_model = converter.convert()
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\lite.py", line 739, in convert
    result = _convert_saved_model(**converter_kwargs)
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 632, in convert_saved_model
    data = toco_convert_protos(
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 216, in toco_convert_protos
    raise ConverterError(str(e))
tensorflow.lite.python.convert.ConverterError: <unknown>:0: error: loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): 'tf.Size' op is neither a custom op nor a flex op
<unknown>:0: note: loc("StatefulPartitionedCall"): called from
<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
```

I have read more issues regarding Flex Ops, unfortunately I haven't found a suitable solution as of now. I tried adding the following line:

```
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
```

When this line is present, the model converts very smoothly to a TFlite model, but then the TFLite model doesn't work anymore on a Raspberry Pi, with the following error: 

```
Unexpected failure when preparing tensor allocations: Regular TensorFlow ops are not supported by this interpreter. Make sure you apply/link the Flex delegate before inference.
```

I never had this before for any other model. The model works perfectly fine with regular TensorFlow. It's just the TFlite part which causes issues for me.

I am using a pre-trained model (SSD MobileNet V2 FPNLite 320x320) from the Detection Model Zoo at https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md

### Source code / logs
I started off by installing a fresh install of Tensorflow. See:

```
(base) C:\Users\Reno>conda activate tensorflow
 
(tensorflow) C:\Users\Reno>pip cache purge
ERROR: No matching packages
 
(tensorflow) C:\Users\Reno>pip list
Package      Version
------------ -------------------
certifi      2020.12.5
pip          20.3.3
setuptools   51.0.0.post20201207
wheel        0.36.2
wincertstore 0.2
 
(tensorflow) C:\Users\Reno>pip install --upgrade tensorflow
Collecting tensorflow
  Downloading tensorflow-2.4.0-cp38-cp38-win_amd64.whl (370.7 MB)
     |████████████████████████████████| 370.7 MB 21 kB/s
Requirement already satisfied: wheel~=0.35 in c:\users\reno\.conda\envs\tensorflow\lib\site-packages (from tensorflow) (0.36.2)
Collecting gast==0.3.3
  Downloading gast-0.3.3-py2.py3-none-any.whl (9.7 kB)
Collecting absl-py~=0.10
  Downloading absl_py-0.11.0-py3-none-any.whl (127 kB)
     |████████████████████████████████| 127 kB 2.2 MB/s
Collecting astunparse~=1.6.3
  Downloading astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting flatbuffers~=1.12.0
  Downloading flatbuffers-1.12-py2.py3-none-any.whl (15 kB)
Collecting google-pasta~=0.2
  Downloading google_pasta-0.2.0-py3-none-any.whl (57 kB)
     |████████████████████████████████| 57 kB 1.5 MB/s
Collecting grpcio~=1.32.0
  Downloading grpcio-1.32.0-cp38-cp38-win_amd64.whl (2.6 MB)
     |████████████████████████████████| 2.6 MB 2.2 MB/s
Collecting h5py~=2.10.0
  Downloading h5py-2.10.0-cp38-cp38-win_amd64.whl (2.5 MB)
     |████████████████████████████████| 2.5 MB 803 kB/s
Collecting keras-preprocessing~=1.1.2
  Downloading Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
     |████████████████████████████████| 42 kB 3.2 MB/s
Collecting numpy~=1.19.2
  Downloading numpy-1.19.5-cp38-cp38-win_amd64.whl (13.3 MB)
     |████████████████████████████████| 13.3 MB 3.3 MB/s
Collecting opt-einsum~=3.3.0
  Downloading opt_einsum-3.3.0-py3-none-any.whl (65 kB)
     |████████████████████████████████| 65 kB 2.2 MB/s
Collecting protobuf>=3.9.2
  Downloading protobuf-3.14.0-py2.py3-none-any.whl (173 kB)
     |████████████████████████████████| 173 kB 2.2 MB/s
Collecting six~=1.15.0
  Downloading six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting tensorboard~=2.4
  Downloading tensorboard-2.4.0-py3-none-any.whl (10.6 MB)
     |████████████████████████████████| 10.6 MB 2.2 MB/s
Requirement already satisfied: setuptools>=41.0.0 in c:\users\reno\.conda\envs\tensorflow\lib\site-packages (from tensorboard~=2.4->tensorflow) (51.0.0.post20201207)
Collecting google-auth<2,>=1.6.3
  Downloading google_auth-1.24.0-py2.py3-none-any.whl (114 kB)
     |████████████████████████████████| 114 kB 2.2 MB/s
Collecting cachetools<5.0,>=2.0.0
  Downloading cachetools-4.2.0-py3-none-any.whl (12 kB)
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Downloading google_auth_oauthlib-0.4.2-py2.py3-none-any.whl (18 kB)
Collecting markdown>=2.6.8
  Downloading Markdown-3.3.3-py3-none-any.whl (96 kB)
     |████████████████████████████████| 96 kB 3.2 MB/s
Collecting pyasn1-modules>=0.2.1
  Downloading pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
     |████████████████████████████████| 155 kB 2.2 MB/s
Collecting pyasn1<0.5.0,>=0.4.6
  Downloading pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
     |████████████████████████████████| 77 kB 2.6 MB/s
Collecting requests<3,>=2.21.0
  Downloading requests-2.25.1-py2.py3-none-any.whl (61 kB)
     |████████████████████████████████| 61 kB 2.0 MB/s
Requirement already satisfied: certifi>=2017.4.17 in c:\users\reno\.conda\envs\tensorflow\lib\site-packages (from requests<3,>=2.21.0->tensorboard~=2.4->tensorflow) (2020.12.5)
Collecting chardet<5,>=3.0.2
  Downloading chardet-4.0.0-py2.py3-none-any.whl (178 kB)
     |████████████████████████████████| 178 kB 2.2 MB/s
Collecting idna<3,>=2.5
  Downloading idna-2.10-py2.py3-none-any.whl (58 kB)
     |████████████████████████████████| 58 kB 2.0 MB/s
Collecting requests-oauthlib>=0.7.0
  Downloading requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
Collecting oauthlib>=3.0.0
  Downloading oauthlib-3.1.0-py2.py3-none-any.whl (147 kB)
     |████████████████████████████████| 147 kB 2.2 MB/s
Collecting rsa<5,>=3.1.4
  Downloading rsa-4.6-py3-none-any.whl (47 kB)
     |████████████████████████████████| 47 kB 3.2 MB/s
Collecting tensorboard-plugin-wit>=1.6.0
  Downloading tensorboard_plugin_wit-1.7.0-py3-none-any.whl (779 kB)
     |████████████████████████████████| 779 kB 2.2 MB/s
Collecting tensorflow-estimator<2.5.0,>=2.4.0rc0
  Downloading tensorflow_estimator-2.4.0-py2.py3-none-any.whl (462 kB)
     |████████████████████████████████| 462 kB 2.2 MB/s
Collecting termcolor~=1.1.0
  Downloading termcolor-1.1.0.tar.gz (3.9 kB)
Collecting typing-extensions~=3.7.4
  Downloading typing_extensions-3.7.4.3-py3-none-any.whl (22 kB)
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.2-py2.py3-none-any.whl (136 kB)
     |████████████████████████████████| 136 kB 2.2 MB/s
Collecting werkzeug>=0.11.15
  Downloading Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
     |████████████████████████████████| 298 kB 2.2 MB/s
Collecting wrapt~=1.12.1
  Downloading wrapt-1.12.1.tar.gz (27 kB)
Building wheels for collected packages: termcolor, wrapt
  Building wheel for termcolor (setup.py) ... done
  Created wheel for termcolor: filename=termcolor-1.1.0-py3-none-any.whl size=4829 sha256=841391132c29825394e6b0ab2a1e79ffb1f6a1867cd2a214dc65cc3c4aa5b688
  Stored in directory: c:\users\reno\appdata\local\pip\cache\wheels\a0\16\9c\5473df82468f958445479c59e784896fa24f4a5fc024b0f501
  Building wheel for wrapt (setup.py) ... done
  Created wheel for wrapt: filename=wrapt-1.12.1-cp38-cp38-win_amd64.whl size=33672 sha256=7efbaa96078fc7ecd811c170d66cc642ab8fd8f88a41af40ba118c5b7c16765f
  Stored in directory: c:\users\reno\appdata\local\pip\cache\wheels\5f\fd\9e\b6cf5890494cb8ef0b5eaff72e5d55a70fb56316007d6dfe73
Successfully built termcolor wrapt
Installing collected packages: urllib3, pyasn1, idna, chardet, six, rsa, requests, pyasn1-modules, oauthlib, cachetools, requests-oauthlib, google-auth, werkzeug, tensorboard-plugin-wit, protobuf, numpy, markdown, grpcio, google-auth-oauthlib, absl-py, wrapt, typing-extensions, termcolor, tensorflow-estimator, tensorboard, opt-einsum, keras-preprocessing, h5py, google-pasta, gast, flatbuffers, astunparse, tensorflow
Successfully installed absl-py-0.11.0 astunparse-1.6.3 cachetools-4.2.0 chardet-4.0.0 flatbuffers-1.12 gast-0.3.3 google-auth-1.24.0 google-auth-oauthlib-0.4.2 google-pasta-0.2.0 grpcio-1.32.0 h5py-2.10.0 idna-2.10 keras-preprocessing-1.1.2 markdown-3.3.3 numpy-1.19.5 oauthlib-3.1.0 opt-einsum-3.3.0 protobuf-3.14.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.25.1 requests-oauthlib-1.3.0 rsa-4.6 six-1.15.0 tensorboard-2.4.0 tensorboard-plugin-wit-1.7.0 tensorflow-2.4.0 tensorflow-estimator-2.4.0 termcolor-1.1.0 typing-extensions-3.7.4.3 urllib3-1.26.2 werkzeug-1.0.1 wrapt-1.12.1

(tensorflow) C:\Users\Reno>cd C:\Users\Reno\Documents\TensorFlow\workspace\training

(tensorflow) C:\Users\Reno\Documents\TensorFlow\workspace\training>python convert_to_tflite_v2.py

2021-01-08 11:30:11.595435: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2021-01-08 11:30:19.711062: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-01-08 11:30:19.715933: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library nvcuda.dll
2021-01-08 11:30:20.065510: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P2000 computeCapability: 6.1
coreClock: 1.607GHz coreCount: 6 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 89.53GiB/s
2021-01-08 11:30:20.072148: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2021-01-08 11:30:20.083133: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2021-01-08 11:30:20.088475: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
2021-01-08 11:30:20.099554: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2021-01-08 11:30:20.105734: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2021-01-08 11:30:20.118062: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2021-01-08 11:30:20.124535: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusparse64_11.dll
2021-01-08 11:30:20.130482: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudnn64_8.dll
2021-01-08 11:30:20.133532: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2021-01-08 11:30:20.136101: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-01-08 11:30:20.144289: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P2000 computeCapability: 6.1
coreClock: 1.607GHz coreCount: 6 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 89.53GiB/s
2021-01-08 11:30:20.152136: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2021-01-08 11:30:20.155789: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2021-01-08 11:30:20.159025: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
2021-01-08 11:30:20.162087: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2021-01-08 11:30:20.165806: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2021-01-08 11:30:20.171067: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2021-01-08 11:30:20.174766: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusparse64_11.dll
2021-01-08 11:30:20.180898: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudnn64_8.dll
2021-01-08 11:30:20.185452: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2021-01-08 11:30:20.767230: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-01-08 11:30:20.771172: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0
2021-01-08 11:30:20.773087: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N
2021-01-08 11:30:20.775197: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2982 MB memory) -> physical GPU (device: 0, name: Quadro P2000, pci bus id: 0000:01:00.0, compute capability: 6.1)
2021-01-08 11:30:20.783136: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-01-08 11:31:15.040028: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:316] Ignored output_format.
2021-01-08 11:31:15.046343: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:319] Ignored drop_control_dependency.
2021-01-08 11:31:15.052724: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:325] Ignored change_concat_input_ranges.
2021-01-08 11:31:15.059159: I tensorflow/cc/saved_model/reader.cc:32] Reading SavedModel from: C:/Users/Reno/Documents/TensorFlow/workspace/training_walnoot/exported-models/walnoot_model_lite/saved_model/
2021-01-08 11:31:15.152959: I tensorflow/cc/saved_model/reader.cc:55] Reading meta graph with tags { serve }
2021-01-08 11:31:15.155845: I tensorflow/cc/saved_model/reader.cc:93] Reading SavedModel debug info (if present) from: C:/Users/Reno/Documents/TensorFlow/workspace/training_walnoot/exported-models/walnoot_model_lite/saved_model/
2021-01-08 11:31:15.163162: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-01-08 11:31:15.167694: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]
2021-01-08 11:31:15.170225: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-01-08 11:31:15.517776: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
2021-01-08 11:31:15.565923: I tensorflow/cc/saved_model/loader.cc:206] Restoring SavedModel bundle.
2021-01-08 11:31:16.326616: I tensorflow/cc/saved_model/loader.cc:190] Running initialization op on SavedModel bundle at path: C:/Users/Reno/Documents/TensorFlow/workspace/training_walnoot/exported-models/walnoot_model_lite/saved_model/
2021-01-08 11:31:16.703255: I tensorflow/cc/saved_model/loader.cc:277] SavedModel load for tags { serve }; Status: success: OK. Took 1644097 microseconds.
2021-01-08 11:31:20.471768: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:194] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2021-01-08 11:31:21.438534: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: Quadro P2000 computeCapability: 6.1
coreClock: 1.607GHz coreCount: 6 deviceMemorySize: 4.00GiB deviceMemoryBandwidth: 89.53GiB/s
2021-01-08 11:31:21.445277: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
2021-01-08 11:31:21.452203: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublas64_11.dll
2021-01-08 11:31:21.456431: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cublasLt64_11.dll
2021-01-08 11:31:21.461100: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cufft64_10.dll
2021-01-08 11:31:21.467734: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library curand64_10.dll
2021-01-08 11:31:21.473807: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusolver64_10.dll
2021-01-08 11:31:21.479188: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cusparse64_11.dll
2021-01-08 11:31:21.485787: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudnn64_8.dll
2021-01-08 11:31:21.491625: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0
2021-01-08 11:31:21.495347: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1261] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-01-08 11:31:21.499727: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267]      0
2021-01-08 11:31:21.502749: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1280] 0:   N
2021-01-08 11:31:21.505539: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2982 MB memory) -> physical GPU (device: 0, name: Quadro P2000, pci bus id: 0000:01:00.0, compute capability: 6.1)
2021-01-08 11:31:21.513866: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): error: 'tf.Size' op is neither a custom op nor a flex op
error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
Traceback (most recent call last):
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 210, in toco_convert_protos
    model_str = wrap_toco.wrapped_toco_convert(model_flags_str,
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\wrap_toco.py", line 32, in wrapped_toco_convert
    return _pywrap_toco_api.TocoConvert(
Exception: <unknown>:0: error: loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): 'tf.Size' op is neither a custom op nor a flex op
<unknown>:0: note: loc("StatefulPartitionedCall"): called from
<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
 
 
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "convert_to_tflite_v2.py", line 7, in <module>
    tflite_model = converter.convert()
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\lite.py", line 739, in convert
    result = _convert_saved_model(**converter_kwargs)
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 632, in convert_saved_model
    data = toco_convert_protos(
  File "C:\Users\Reno\.conda\envs\tensorflow\lib\site-packages\tensorflow\lite\python\convert.py", line 216, in toco_convert_protos
    raise ConverterError(str(e))
tensorflow.lite.python.convert.ConverterError: <unknown>:0: error: loc(callsite(callsite("Postprocessor/BatchMultiClassNonMaxSuppression/MultiClassNonMaxSuppression/SortByField_1/Size@__inference___call___23519" at "StatefulPartitionedCall@__inference_signature_wrapper_25508") at "StatefulPartitionedCall")): 'tf.Size' op is neither a custom op nor a flex op
<unknown>:0: note: loc("StatefulPartitionedCall"): called from
<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag):
        tf.Size {device = ""}
 
 
(tensorflow) C:\Users\Reno\Documents\TensorFlow\workspace\training>
```

Why is this error appearing just now, what does it mean, and how can I convert a model to TFlite without any of these nasty errors and subsequently run it on a Raspberry Pi without any hassle.

Thanks in advance.

### PR DESCRIPTION
**Pull request by WindQAQ**
This PR converts `TF_SizeOp` to `TF_ProdOp(TF_ShapeOp)`. It makes TFL users can run models with `tf.size` without `SELECT_TF_OPS`. Fixes #46285.

### CODE REVIEW

**Review by abattery — COMMENTED** [review]
Could you move this pattern into lower_tf.cc under tensorflow/compiler/mlir/tensorflow/transforms?

**Comment by WindQAQ** [comment]
Done. Moved the pattern to tf-lower-tf.

File: tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td [hunk 1]
```diff
@@ -252,6 +258,17 @@ def : Pat<(TF_RsqrtGradOp $lhs, $rhs),
                               (TF_ConstOp (GetScalarOfType<-2> $rhs))))>;
 
 
+//===----------------------------------------------------------------------===//
+// Size op patterns.
+//===----------------------------------------------------------------------===//
+
+// Size(x) = Prod(Shape(x), axis=0, keepdims=false)
+def : Pat<(TF_SizeOp:$result $arg),
+          (TF_ProdOp (CreateTFShapeOp $result, $arg, (IsI32 $result)),
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:267]: /*reduction_indices=*/(GetScalarOfType<0> $result)

As reduction_indices in prod to be of i32 if the elements are i32 and I think scalar reduction indices are allowed.

File: tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td [hunk 2]
```diff
@@ -252,6 +258,17 @@ def : Pat<(TF_RsqrtGradOp $lhs, $rhs),
...
+                     (TF_ConstOp (GetI64ScalarElementsAttr<0>)),
+                     ConstBoolAttrFalse)>;
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.td:269]: nit: /*keep_dims=*/ConstBoolAttrFalse for readability.

**Review by smit-hinsu — COMMENTED** [review]
Thanks for the contributions!

**Comment by WindQAQ** [comment]
Hi @smit-hinsu, TF->TF lowering pass is also being used in `tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc`, so this PR will break `size_unranked` in `tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir`. Should I remove it or correct it to the right lowering? Thank you!

https://github.com/tensorflow/tensorflow/blob/009675faa9422ef8e2c8a30216b31adad62d25b6/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir#L3882-L3887

**Review by smit-hinsu — COMMENTED** [review]
I will need to take a closer look but this will also change lowering for the static shaped input which I think should be fine.

Could you try removing the lowering pattern as well?

**Comment by WindQAQ** [comment]
Sure, I can constrain the pattern to ranked input only.

**Review by smit-hinsu — COMMENTED** [review]
Why not simply remove that pattern? Your new pattern should have the same behavior in case of static shapes.

**Comment by WindQAQ** [comment]
> Why not simply remove that pattern? Your new pattern should have the same behavior in case of static shapes.

Ohoh, I misunderstand it :-(. Will do. Thank you!

**Comment by WindQAQ** [comment]
Umm, sorry that I mess up the branch. Let me reopen a new one.

**Pull request closed without merging**